File: dailyQuestion/2020/2020-09/9-23/golang/solution_dp.go
package main

import (
	"fmt"
	"math"
)

func main() {
	result := minimumOperations("rrryyyrryyyrr")
	fmt.Println(result)
}

const inf = math.MaxInt32

func minimumOperations(leaves string) int {
	n := len(leaves)
	f := make([][3]int, n)
	f[0][0] = boolToInt(leaves[0] == 'y')
	f[0][1] = inf
	f[0][2] = inf
	f[1][2] = inf
	for i := 1; i < n; i++ {
		f[i][0] = f[i-1][0] + boolToInt(leaves[i] == 'y')
		f[i][1] = min(f[i-1][0], f[i-1][1]) + boolToInt(leaves[i] == 'r')
		if i >= 2 {
			f[i][2] = min(f[i-1][1], f[i-1][2]) + boolToInt(leaves[i] == 'y')
		}
	}
	return f[n-1][2]
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

func boolToInt(b bool) int {
	if b {
		return 1
	}
	return 0
}

File: dailyQuestion/2020/2020-09/9-23/python/solution_dp.py
def minimumOperations(leaves: str) -> int:
    def boolToInt(b) :
        if b :
            return 1
        return 0

    n = len(leaves)
    f = [ [float('inf')] * 3  for i in range(n)]
    f[0][0] = boolToInt(leaves[0]=='y')
    for i in range(1,n) :
        f[i][0] = f[i-1][0] + boolToInt(leaves[i]=='y')
        f[i][1] = min(f[i-1][0], f[i-1][1]) + boolToInt(leaves[i]=='r')
        if i >= 2 :
            f[i][2] = min(f[i-1][1], f[i-1][2]) + boolToInt(leaves[i] == 'y')
    return f[n-1][2]


if __name__ == "__main__" :
    result = minimumOperations("rrryyyrryyyrr")
    print(result)

